File: androidApp/v2/core/database/src/main/java/com/deadly/v2/core/database/dao/ShowDao.kt
package com.deadly.v2.core.database.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import com.deadly.v2.core.database.entities.ShowEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface ShowDao {
    
    // Core operations for import
    @Insert
    suspend fun insert(show: ShowEntity)
    
    @Insert
    suspend fun insertAll(shows: List<ShowEntity>)
    
    // Basic queries for verification
    @Query("SELECT * FROM shows ORDER BY date DESC")
    suspend fun getAllShows(): List<ShowEntity>
    
    @Query("SELECT * FROM shows ORDER BY date DESC")
    fun getAllShowsFlow(): Flow<List<ShowEntity>>
    
    @Query("SELECT * FROM shows WHERE showId = :showId")
    suspend fun getShowById(showId: String): ShowEntity?
    
    @Query("SELECT * FROM shows WHERE showId IN (:showIds)")
    suspend fun getShowsByIds(showIds: List<String>): List<ShowEntity>
    
    @Query("SELECT COUNT(*) FROM shows")
    suspend fun getShowCount(): Int
    
    // Date-based queries
    @Query("SELECT * FROM shows WHERE year = :year ORDER BY date")
    suspend fun getShowsByYear(year: Int): List<ShowEntity>
    
    @Query("SELECT * FROM shows WHERE yearMonth = :yearMonth ORDER BY date")
    suspend fun getShowsByYearMonth(yearMonth: String): List<ShowEntity>
    
    @Query("SELECT * FROM shows WHERE date = :date ORDER BY showSequence")
    suspend fun getShowsByDate(date: String): List<ShowEntity>
    
    @Query("SELECT * FROM shows WHERE date >= :startDate AND date <= :endDate ORDER BY date")
    suspend fun getShowsInDateRange(startDate: String, endDate: String): List<ShowEntity>
    
    
    // Location queries
    @Query("SELECT * FROM shows WHERE venueName LIKE '%' || :venueName || '%' ORDER BY date")
    suspend fun getShowsByVenue(venueName: String): List<ShowEntity>
    
    @Query("SELECT * FROM shows WHERE city = :city ORDER BY date DESC")
    suspend fun getShowsByCity(city: String): List<ShowEntity>
    
    @Query("SELECT * FROM shows WHERE state = :state ORDER BY date DESC")
    suspend fun getShowsByState(state: String): List<ShowEntity>
    
    // Search queries
    @Query("""
        SELECT * FROM shows 
        WHERE songList LIKE '%' || :songName || '%' 
        ORDER BY date DESC
    """)
    suspend fun getShowsBySong(songName: String): List<ShowEntity>
    
    // Popular/featured queries
    @Query("SELECT * FROM shows WHERE averageRating IS NOT NULL ORDER BY averageRating DESC LIMIT :limit")
    suspend fun getTopRatedShows(limit: Int = 20): List<ShowEntity>
    
    @Query("SELECT * FROM shows ORDER BY date DESC LIMIT :limit")
    suspend fun getRecentShows(limit: Int = 20): List<ShowEntity>
    
    @Query("SELECT * FROM shows WHERE month = :month AND SUBSTR(date, 9, 2) = PRINTF('%02d', :day) AND recordingCount > 0 ORDER BY year")
    suspend fun getShowsForDate(month: Int, day: Int): List<ShowEntity>
    
    // Navigation queries for efficient chronological traversal
    @Query("SELECT * FROM shows WHERE date > :currentDate ORDER BY date ASC LIMIT 1")
    suspend fun getNextShowByDate(currentDate: String): ShowEntity?
    
    @Query("SELECT * FROM shows WHERE date < :currentDate ORDER BY date DESC LIMIT 1") 
    suspend fun getPreviousShowByDate(currentDate: String): ShowEntity?
    
    // Specific famous shows for verification
    @Query("SELECT * FROM shows WHERE date = '1977-05-08'")
    suspend fun getCornell77(): List<ShowEntity>
    
    // Management operations
    @Query("DELETE FROM shows")
    suspend fun deleteAll()
}

File: iosApp/deadly/Core/Database/DAOs/ShowDAO.swift
import GRDB

struct ShowDAO: Sendable {
    let database: AppDatabase

    // MARK: - Insert

    func insert(_ show: ShowRecord) throws {
        try database.write { db in
            var record = show
            try record.insert(db)
        }
    }

    func insertAll(_ shows: [ShowRecord]) throws {
        try database.write { db in
            for show in shows {
                var record = show
                try record.insert(db)
            }
        }
    }

    // MARK: - Fetch

    func fetchAll() throws -> [ShowRecord] {
        try database.read { db in
            try ShowRecord.order(Column("date").asc).fetchAll(db)
        }
    }

    func fetchById(_ showId: String) throws -> ShowRecord? {
        try database.read { db in
            try ShowRecord.fetchOne(db, key: showId)
        }
    }

    func fetchByIds(_ showIds: [String]) throws -> [ShowRecord] {
        try database.read { db in
            try ShowRecord.fetchAll(db, keys: showIds)
        }
    }

    func fetchCount() throws -> Int {
        try database.read { db in
            try ShowRecord.fetchCount(db)
        }
    }

    // MARK: - Date queries

    func fetchByYear(_ year: Int) throws -> [ShowRecord] {
        try database.read { db in
            try ShowRecord
                .filter(Column("year") == year)
                .order(Column("date").asc)
                .fetchAll(db)
        }
    }

    func fetchByYearMonth(_ yearMonth: String) throws -> [ShowRecord] {
        try database.read { db in
            try ShowRecord
                .filter(Column("yearMonth") == yearMonth)
                .order(Column("date").asc)
                .fetchAll(db)
        }
    }

    func fetchByDate(_ date: String) throws -> [ShowRecord] {
        try database.read { db in
            try ShowRecord
                .filter(Column("date") == date)
                .order(Column("showSequence").asc)
                .fetchAll(db)
        }
    }

    func fetchInDateRange(start: String, end: String) throws -> [ShowRecord] {
        try database.read { db in
            try ShowRecord
                .filter(Column("date") >= start && Column("date") <= end)
                .order(Column("date").asc)
                .fetchAll(db)
        }
    }

    // MARK: - Location queries

    func fetchByVenue(_ venueName: String) throws -> [ShowRecord] {
        let pattern = "%\(escapeLike(venueName))%"
        return try database.read { db in
            try ShowRecord
                .filter(sql: "venueName LIKE ? ESCAPE '\\'", arguments: [pattern])
                .order(Column("date").asc)
                .fetchAll(db)
        }
    }

    func fetchByCity(_ city: String) throws -> [ShowRecord] {
        let pattern = "%\(escapeLike(city))%"
        return try database.read { db in
            try ShowRecord
                .filter(sql: "city LIKE ? ESCAPE '\\'", arguments: [pattern])
                .order(Column("date").asc)
                .fetchAll(db)
        }
    }

    func fetchByState(_ state: String) throws -> [ShowRecord] {
        try database.read { db in
            try ShowRecord
                .filter(Column("state") == state)
                .order(Column("date").asc)
                .fetchAll(db)
        }
    }

    // MARK: - Popular / featured

    func fetchTopRated(limit: Int) throws -> [ShowRecord] {
        try database.read { db in
            try ShowRecord
                .filter(Column("averageRating") != nil)
                .order(Column("averageRating").desc, Column("totalReviews").desc)
                .limit(limit)
                .fetchAll(db)
        }
    }

    /// Returns shows that occurred on the same month/day in any year (anniversary).
    func fetchOnThisDay(month: Int, day: Int) throws -> [ShowRecord] {
        try database.read { db in
            try ShowRecord
                .filter(sql: "month = ? AND CAST(SUBSTR(date, 9, 2) AS INTEGER) = ?", arguments: [month, day])
                .order(Column("year").asc)
                .fetchAll(db)
        }
    }

    // MARK: - Chronological navigation

    func fetchNext(after date: String) throws -> ShowRecord? {
        try database.read { db in
            try ShowRecord
                .filter(Column("date") > date)
                .order(Column("date").asc)
                .fetchOne(db)
        }
    }

    func fetchPrevious(before date: String) throws -> ShowRecord? {
        try database.read { db in
            try ShowRecord
                .filter(Column("date") < date)
                .order(Column("date").desc)
                .fetchOne(db)
        }
    }

    // MARK: - Search (LIKE fallback for short queries)

    /// LIKE search across venueName, city, state, and date — used when query is ≤2 chars.
    func searchLike(_ query: String) throws -> [ShowRecord] {
        let pattern = "%\(escapeLike(query))%"
        return try database.read { db in
            try ShowRecord
                .filter(sql: """
                    venueName LIKE ? ESCAPE '\\' OR
                    city LIKE ? ESCAPE '\\' OR
                    state LIKE ? ESCAPE '\\' OR
                    date LIKE ? ESCAPE '\\'
                    """, arguments: [pattern, pattern, pattern, pattern])
                .order(Column("date").asc)
                .limit(100)
                .fetchAll(db)
        }
    }

    // MARK: - Management

    func deleteAll() throws {
        try database.write { db in
            try ShowRecord.deleteAll(db)
        }
    }

    // MARK: - Private helpers

    private func escapeLike(_ value: String) -> String {
        value
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "%", with: "\\%")
            .replacingOccurrences(of: "_", with: "\\_")
    }
}
